zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,187 @@
+        let R_i = NonceCommitment(RISTRETTO_BASEPOINT_POINT * r_i);
+        // Make H(R_i)
+        let precommitment = R_i.precommit();
+
+        (
+            PartyAwaitingPrecommitments {
+                transcript: transcript.clone(),
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                nonce_precommitments,
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
+    ) -> Result<(PartyAwaitingSiglets, Scalar), VMError> {
+        // Check stored precommitments against received commitments
+        for (pre_comm, comm) in self
+            .nonce_precommitments
+            .iter()
+            .zip(nonce_commitments.iter())
+        {
+            // Make H(comm) = H(R_i)
+            let actual_precomm = comm.precommit();
+
+            // Compare H(comm) with pre_comm, they should be equal
+            let equal = pre_comm.0.ct_eq(&actual_precomm.0);
+            if equal.unwrap_u8() == 0 {
+                return Err(VMError::InconsistentWitness);
+            }
+        }
+
+        // Make R = sum_i(R_i). nonce_commitments = R_i from all the parties.
+        let R: RistrettoPoint = nonce_commitments.iter().map(|R_i| R_i.0).sum();
+
+        // Make c = H(X, R, m)
+        // The message `m` should already have been fed into the transcript.
+        let c = {
+            self.transcript
+                .commit_point(b"P", &self.multikey.aggregated_key().0);
+            self.transcript.commit_point(b"R", &R.compress());
+            self.transcript.challenge_scalar(b"c")
+        };
+
+        // Make a_i = H(L, X_i)
+        let X_i = VerificationKey((self.x_i * RISTRETTO_BASEPOINT_POINT).compress());
+        let a_i = self.multikey.factor_for_key(&X_i);
+
+        // Generate share: s_i = r_i + c * a_i * x_i
+        let s_i = self.r_i + c * a_i * self.x_i;
+
+        // Store received nonce commitments in next state
+        Ok((
+            PartyAwaitingSiglets {
+                multikey: self.multikey,
+                nonce_commitments,
+                c,
+                R,
+            },
+            s_i,
+        ))
+    }
+}
+
+impl PartyAwaitingSiglets {
+    pub fn receive_trusted_shares(self, shares: Vec<Scalar>) -> Signature {
+        // s = sum(s_i), s_i = shares[i]
+        let s: Scalar = shares.iter().map(|share| share).sum();
```

Comment:
maybe moving things in (since we don't reuse them later) will be more efficient?
```suggestion
        let s: Scalar = shares.into_iter().map(|share| share).sum();
```

Reply:
Why would this be more efficient... because the function would have ownership of the shares?

I made the change!